### DISCUSSION TITLE
Purpose of mocks contracts in Tests

### DISCUSSION THREAD

**Discussion by Shawon-Mandal** [Cyfrin/security-and-auditing-full-course-s23#210] [Q&A] [answered]
Currently I am doing the Thunder Loan part of the course and I am very new about writing tests and all about mocks.
This might sound like a stupid question, but I want to know what are the purpose of mocks contracts in tests.

Inside the unit tests of Thunderloan, I see that many of the contracts imports mocks or interfaces to write tests instead of importing the original contracts. I wonder why, can anyone please explain it to me?
![Thunderloan](https://github.com/user-attachments/assets/b7cee042-5621-4aac-8a5b-045a70993655)
![Thunderloan2](https://github.com/user-attachments/assets/02b7fc79-12b4-404f-a313-da2ba9aa3713)

If I want to set up a foundry testing environment from scratch how should I set up everything?
Particularly, which contracts should I import in the `contractname.t.sol` to properly set everything up?

Thank you

**Comment by EngrPips** [comment]
Hello @Shawon-Mandal, We use Mocks when we want to have a stripped-down version of the original contract to be able to test on local chains such as `Anvil` or any other local chain you are working with. If you want to set up a foundry testing environment from scratch, what you are testing will determine what you will set up and how you will set it up.

**Reply by Shawon-Mandal** [reply]
Hello @EngrPips, Thank you for responding. Let's say, I want to write a test for a contract which has a library but that library does not include all the functions of the original contract. Now, How should I set it up? 

Also, Is it a normal thing for auditors to use mock contracts to write tests on? 
I am asking this because, wouldn't I miss potential bug if I use a stripped down version of that particular smart contract?

**Reply by EngrPips** [reply, marked as answer]
Why does the library not include all the functions of the original contract? You only use mock when you are sure it is best to use, and when you need to test the real contract and don't want to miss any details, you will use test net like sepolia.

**Reply by Shawon-Mandal** [reply]
I think I got the idea. Thank you

**Reply by EngrPips** [reply]
You are welcome. Keep crushing it.
